### PR TITLE
fix(widgets): reduces BC break possibility in widgets JS code

### LINKED ISSUE #9523: Possible BC break in new widgets AMD module
See https://github.com/ColdTrick/widget_manager/issues/76

@hypeJunction details would be great here.

### PR DESCRIPTION
**Pull request by hypeJunction**
Now calls the init,system event to make the behavior similar to 2.0

Fixes #9523

### CODE REVIEW

**Comment by hypeJunction** [comment]
@mrclay This issues a deprecation message on all methods of `elgg.ui.widgets`, which is I think is the bigger problem than init.

File: js/lib/ui.widgets.js [hunk 1]
```diff
@@ -1,223 +1,35 @@
-elgg.provide('elgg.ui.widgets');
+require(['elgg', 'jquery', 'elgg/widgets'], function (elgg, $, widgets) {
```
Comment (mrclay) [inline js/lib/ui.widgets.js]: This works as long as we inline elgg/widgets and this async function runs immediately. That's probably safe for BC and the cleanup is definitely easier later.

File: js/lib/ui.widgets.js [hunk 2]
```diff
@@ -1,223 +1,35 @@
...
-};
-
-require(['jquery'], function ($) {
-	$(function () {
-		require(['elgg/widgets'], function (widgets) {
-			widgets.init();
-		});
-	});
-});
+	elgg.register_hook_handler('init', 'system', widgets.init);
```
Comment (mrclay) [inline js/lib/ui.widgets.js]: This should probably go in the widget.

File: views/default/elgg.js.php [hunk 3]
```diff
@@ -18,6 +18,8 @@
 // For backwards compatibility...
 echo elgg_view('sprintf.js');
 
+echo elgg_view('elgg/widgets.js');
```
Comment (mrclay) [inline views/default/elgg.js.php]: Need a comment before this, like "Inlining this is required for the elgg.ui.widgets API to be defined immediately. Without this, 3rd party plugins calling elgg.ui.widget methods will break."

File: views/default/elgg/widgets.js [hunk 4]
```diff
@@ -1,13 +1,208 @@
...
+	 *
+	 * @param {Object} event
+	 * @return void
+	 */
+	widgets.collapseToggle = function (event) {
+		$(this).toggleClass('elgg-widget-collapsed');
+		$(this).parent().parent().find('.elgg-body').slideToggle('medium');
+		event.preventDefault();
+	},
+			/**
```
Comment (mrclay) [inline views/default/elgg/widgets.js]: indent

**Comment by mrclay** [comment]
This solution makes me nervous, but the cleanliness is a big win and it's probably fine.

**Comment by mrclay** [comment]
Commit description maybe should mention that this inlines elgg/widget and relies on "almost synchronous" behavior of `require` since the 3 dependents are already loaded.

File: js/lib/ui.widgets.js (hunk 2 above)
Comment (hypeJunction) [inline js/lib/ui.widgets.js]: I want to remove it in 3.0 and make elgg/widgets require elgg/ready

**Comment by hypeJunction** [comment]
@mrclay Updated commit message and added inline docs

File: views/default/elgg/widgets.js [hunk 5]
```diff
@@ -1,13 +1,223 @@
...
-define(function (require) {
-	var elgg = require('elgg');
-	var $ = require('jquery');
+/**
+ * To implement a BC solution and deprecated elgg.ui.widgets, we use a named
+ * AMD module that is inlined in elgg.js. This allows us to proxy elgg.ui.widgets
+ * calls via elgg/widgets module to issue deprecation notices.
+ *
+ * @todo: in 3.x make this an anonymous module
+ * @todo: in 3.x define elgg/ready as a dependency
```
Comment (mrclay) [inline views/default/elgg/widgets.js]: Shouldn't it be elgg/init?

Comment (mrclay) [inline views/default/elgg/widgets.js]: Never mind. As long as plugins depend on the module it doesn't matter.

**Comment by mrclay** [comment]
LGTM

**Comment by hypeJunction** [comment]
Can we get another LGTM and merge?

**Pull request merged**